pkgs/development/tools/jdt-language-server/default.nix

```diff
@@ -0,0 +1,99 @@
+{ lib
+, stdenv
+, fetchurl
+, makeWrapper
+, jdk15
+}:
+
+stdenv.mkDerivation rec {
+  pname = "jdt-language-server";
+  version = "0.67.0";
+  timestamp = "202012170459";
+
+  src = fetchurl {
+    url = "https://download.eclipse.org/jdtls/milestones/${version}/jdt-language-server-${version}-${timestamp}.tar.gz";
+    sha256 = "11qzwf2l737hdv1g7bfcrmdxcrl7g9z2k9nhy9lknlabmhq4mcdl";
+  };
+
+  sourceRoot = ".";
+
+  buildInputs = [
+    jdk15
+  ];
+
+  nativeBuildInputs = [
+    makeWrapper
+  ];
+
+  installPhase =
+    let
+      # The application ships with config directories for linux and mac
+      configDir = if stdenv.isDarwin then "config_mac" else "config_linux";
+      # The application will store it's data here. Note especially the escaping so
+      # that the env vars are interpreted at runtime and not during installPhase.
+      runtimePath = "\\\${XDG_CACHE_HOME:-\$HOME/.cache}/jdt-language-server";
+    in
+    ''
+      # Copy jars
+      install -D -t $out/share/java/plugins/ plugins/*.jar
+
+      # Copy config directories for linux and mac
+      install -Dm 444 -t $out/share/config ${configDir}/*
+
+      # Get latest version of launcher jar
+      # e.g. org.eclipse.equinox.launcher_1.5.800.v20200727-1323.jar
+      launcher="$(ls $out/share/java/plugins/org.eclipse.equinox.launcher_* | sort -V | tail -n1)"
+
+      # The wrapper script will create a directory in the user's cache, copy in the config
+      # files since this dir can't be read-only, and by default use this as the runtime dir.
+      #
+      # The following options are required as per the upstream documentation:
+      #
+      #   -Declipse.application=org.eclipse.jdt.ls.core.id1
+      #   -Dosgi.bundles.defaultStartLevel=4
+      #   -Declipse.product=org.eclipse.jdt.ls.core.product
+      #   -noverify
+      #   --add-modules=ALL-SYSTEM
+      #   --add-opens java.base/java.util=ALL-UNNAMED
+      #   --add-opens java.base/java.lang=ALL-UNNAMED
+      #
+      # Other options which the user may change:
+      #
+      #   -Dlog.level:
+      #     Log level.
+      #     This can be overidden by setting JAVA_OPTS.
+      #   -configuration:
+      #     The application will read the configuration from this directory but also needs
+      #     to be able to write here (hence mode 17777).
+      #     This can be overidden by specifying -configuration to the wrapper.
+      #   -data:
+      #     The application stores runtime data here. We set this to <cache-dir>/$PWD
+      #     so that projects don't collide with each other.
+      #     This can be overidden by specifying -configuration to the wrapper.
+      #
+      # Java options, such as -Xms and Xmx can be specified by setting JAVA_OPTS.
+      makeWrapper ${jdk15}/bin/java $out/bin/jdt-language-server \
```

Comment:
I'm not 100% sure, but wouldn't be better to use java in PATH? That's the one usually used by your project, not sure if there will be some mismatches if we use different versions. In any case, I think it would be a good idea to use one jdk as fallback, in case there it no java in PATH, but why jdk15? Wouldn't be better `jdk`?

Reply:
Do we usually package applications like this in nixpkgs? Is there an example you can point me to? Does `jdk` always track the latest version? If so, I would agree that it makes sense to use that. @fzakaria suggested to switch to a specific version though above.

Reply:
Additionally, a user can always override the JDK using an overlay, which would look something like the below. I didn't test this out so it might not be 100% correct.

```nix
self: super: 
{
  jdt-language-server = super.jdt-language-server.override {
    jdk = jdk9;
  };
}
```

Reply:
> Do we usually package applications like this in nixpkgs? Is there an example you can point me to? Does jdk always track the latest version? If so, I would agree that it makes sense to use that. @fzakaria suggested to switch to a specific version though above.

I think that the general approach here is to use `jdk` in `default.nix` and to point to a specific jdk version in `all-packages.nix`

I asked the same question here:
https://discourse.nixos.org/t/package-version-arguments-vs-generic-package-arguments/9071
this PR introduced that pattern: https://github.com/NixOS/nixpkgs/pull/89731

And here you can see the changes I did for clojure, since it needs to use jdk11:
https://github.com/NixOS/nixpkgs/pull/106462/files